File: 10130 SuperSale/solution.c
#include <stdio.h>

typedef struct {
    short P, W;
} Object;

int main() {
	int T;
	scanf("%d", &T);
	while(T--) {
	    int N;
	    scanf("%d", &N);
	    Object objects[N];
	    for(int i = 0; i < N; ++i)
	        scanf("%hd %hd", &objects[i].P, &objects[i].W);
	    int G;
	    scanf("%d", &G);
	    short MW[G], capacity = 0;
	    for(int i = 0; i < G; ++i) {
	        scanf("%hd", &MW[i]);
	        capacity = capacity > MW[i] ? capacity : MW[i];
	    }
	    int dp[capacity + 1];
	    memset(dp, 0, sizeof(dp));
	    for(int i = 0; i < N; ++i) {
	        short p = objects[i].P;
	        short w = objects[i].W;
	        for(int j = capacity; j >= w; --j)
	            dp[j] = dp[j] > dp[j - w] + p ? dp[j] : dp[j - w] + p;
	    }
	    int maxValue = 0;
	    for(int i = 0; i < G; ++i)
	        maxValue += dp[MW[i]];
	    printf("%d\n", maxValue);
	}
	return 0;
}

File: 10130 SuperSale/solution.java
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
        st.nextToken();
        int T = (int)st.nval;
        StringBuilder output = new StringBuilder();
        while(T-- > 0) {
            st.nextToken();
            int N = (int)st.nval;
            Object[] objects = new Object[N];
            for(int i = 0; i < N; ++i) {
                objects[i] = new Object();
                st.nextToken();
                objects[i].P = (short)st.nval;
                st.nextToken();
                objects[i].W = (short)st.nval;
            }
            st.nextToken();
            int G = (int)st.nval;
            short[] MW = new short[G];
            short capacity = 0;
            for(int i = 0; i < G; ++i) {
                st.nextToken();
                capacity = (short)Math.max(capacity, MW[i] = (short)st.nval);
            }
            int[] dp = new int[capacity + 1];
            for(Object o : objects) {
                for(int j = capacity; j >= o.W; --j)
                    dp[j] = Math.max(dp[j], dp[j - o.W] + o.P);
            }
            int maxValue = 0;
            for(Short s : MW)
                maxValue += dp[s];
            output.append(maxValue).append('\n');
    	}
        System.out.print(output);
	}
}

class Object {
    short P, W;
}
